File: Backend/ProjectK.Backend/ProjectK.BusinessLogic/Modules/KurinModule/Features/Registry/Export/RegistryColumns.cs
using ProjectK.BusinessLogic.Modules.KurinModule.Models;
using ProjectK.Common.Interfaces.Modules.InfrastructureModule;
using ProjectK.Common.Models.Enums;

namespace ProjectK.BusinessLogic.Modules.KurinModule.Features.Registry.Export;

/// <summary>
/// What each registry column id means. The ids are the frontend's — the screen owns which columns
/// exist and what they are called, and this is the same list on the writing side.
/// <para>
/// Кожен ступінь — колонка <c>level:&lt;PlastLevel&gt;</c>, тож новий ступінь з'являється тут сам,
/// щойно потрапляє в enum. Іменовані колонки доводиться додавати руками; колонка, якої тут немає,
/// просто не вивантажується.
/// </para>
/// <para>
/// Підписи ступенів беруться з <see cref="PlastLevelNames"/> — того самого місця, звідки їх бере
/// звіт куреня. Свою таблицю тут тримати не можна: два підписи одного ступеня в одному релізі
/// вже траплялися.
/// </para>
/// </summary>
internal static class RegistryColumns
{
    private sealed record Column(string Header, Func<MemberResponse, SheetCell> Read);

    private static readonly IReadOnlyDictionary<string, Column> Named =
        new Dictionary<string, Column>(StringComparer.Ordinal)
        {
            ["dateOfBirth"] = new("Дата народження", m => SheetCell.Of(m.DateOfBirth)),
            ["plastLevel"] = new("Пластовий ступінь", m => SheetCell.Of(LevelName(m.LatestPlastLevel))),
            ["groupName"] = new("Гурток", m => SheetCell.Of(m.GroupName)),
            ["mentoredGroups"] = new("Гурток (закріплення)", m => SheetCell.Of(
                m.MentoredGroupNames.Count == 0 ? null : string.Join(", ", m.MentoredGroupNames))),
            ["phoneNumber"] = new("Телефон", m => SheetCell.Of(m.PhoneNumber)),
            ["email"] = new("Пошта", m => SheetCell.Of(m.Email)),
            ["address"] = new("Адреса", m => SheetCell.Of(m.Address)),
            ["school"] = new("Школа", m => SheetCell.Of(m.School))
        };

    /// <summary>The column's heading, or null when nothing here answers to that id.</summary>
    public static string? HeaderFor(string columnId)
    {
        if (Named.TryGetValue(columnId, out var named))
        {
            return named.Header;
        }

        return TryLevel(columnId, out var level) ? PlastLevelNames.Of(level) : null;
    }

    /// <summary>This person's value in that column.</summary>
    public static SheetCell Read(string columnId, MemberResponse member)
    {
        if (Named.TryGetValue(columnId, out var named))
        {
            return named.Read(member);
        }

        if (!TryLevel(columnId, out var level))
        {
            return SheetCell.Of((string?)null);
        }

        var reached = member.PlastLevelHistories.FirstOrDefault(history => history.PlastLevel == level);
        return SheetCell.Of(reached?.DateAchieved);
    }

    private static bool TryLevel(string columnId, out PlastLevel level)
    {
        level = default;
        const string prefix = "level:";
        return columnId.StartsWith(prefix, StringComparison.Ordinal)
               && Enum.TryParse(columnId[prefix.Length..], out level)
               && Enum.IsDefined(level);
    }

    private static string? LevelName(PlastLevel? level)
        => level.HasValue ? PlastLevelNames.Of(level.Value) : null;
}

File: Frontend/projectk-frontend/src/app/features/kurinModule/pages/registry/registry-columns.ts
import { MemberDto } from '../../models/member.dto';
import { KurinBranch } from '../../models/enums/kurin-branch.enum';
import { PlastLevel } from '../../models/enums/plast-level.enum';
import { PLAST_LADDER, PLAST_LEVEL_COLUMN_LABELS, defaultLevelsFor } from '../../models/enums/plast-ladder';

/** Як малюється клітинка. Дати йдуть через `date`, решта — текстом. */
export type RegistryColumnKind = 'text' | 'date';

export interface RegistryColumn {
  readonly id: string;
  readonly header: string;
  /** Група в перемикачі колонок, щоб одинадцять дат не лежали одним списком. */
  readonly group: 'Особа' | 'Контакти' | 'Належність' | 'Ступені';
  readonly kind: RegistryColumnKind;
  /** Значення для сортування, фільтра й друку. */
  readonly value: (member: MemberDto) => string | Date | null;
}

function levelDate(member: MemberDto, level: PlastLevel): Date | null {
  const entry = (member.plastLevelHistories ?? []).find(history => history.plastLevel === level);
  if (!entry?.dateAchieved) {
    return null;
  }

  return entry.dateAchieved instanceof Date ? entry.dateAchieved : new Date(entry.dateAchieved);
}

const LEVEL_COLUMNS: RegistryColumn[] = PLAST_LADDER.map(level => ({
  id: `level:${level}`,
  header: PLAST_LEVEL_COLUMN_LABELS[level],
  group: 'Ступені' as const,
  kind: 'date' as const,
  value: (member: MemberDto) => levelDate(member, level)
}));

/**
 * Кожна колонка реєстру. Прізвище й ім'я не тут — вони в замороженій першій колонці, яку не
 * вимкнути: рядок без імені нічого не означає.
 */
export const REGISTRY_COLUMNS: RegistryColumn[] = [
  {
    id: 'dateOfBirth',
    header: 'Дата народження',
    group: 'Особа',
    kind: 'date',
    value: member => member.dateOfBirth ?? null
  },
  {
    id: 'plastLevel',
    header: 'Пластовий ступінь',
    group: 'Особа',
    kind: 'text',
    value: member => member.latestPlastLevelDisplay ?? null
  },
  {
    id: 'groupName',
    header: 'Гурток',
    group: 'Належність',
    kind: 'text',
    value: member => member.groupName ?? null
  },
  {
    id: 'mentoredGroups',
    header: 'Гурток (закріплення)',
    group: 'Належність',
    kind: 'text',
    value: member => member.mentoredGroupNames?.join(', ') || null
  },
  {
    id: 'phoneNumber',
    header: 'Телефон',
    group: 'Контакти',
    kind: 'text',
    value: member => member.phoneNumber || null
  },
  {
    id: 'email',
    header: 'Пошта',
    group: 'Контакти',
    kind: 'text',
    value: member => member.email || null
  },
  {
    id: 'address',
    header: 'Адреса',
    group: 'Контакти',
    kind: 'text',
    value: member => member.address ?? null
  },
  {
    id: 'school',
    header: 'Школа',
    group: 'Контакти',
    kind: 'text',
    value: member => member.school ?? null
  },
  ...LEVEL_COLUMNS
];

/**
 * Що курінь цієї гілки бачить, поки нічого не вибирав. Особа й належність — завжди; з дат — те, що
 * гілка звично записує.
 *
 * `mentoredGroups` тут немає навмисно: у таблиці впорядників вона стоїть завжди, а в таблиці юнаків
 * не значить нічого — вмикати її вручну ніде.
 */
export function defaultColumnIdsFor(branch: KurinBranch | null | undefined): string[] {
  return [
    'dateOfBirth',
    'plastLevel',
    'groupName',
    'phoneNumber',
    ...defaultLevelsFor(branch).map(level => `level:${level}`)
  ];
}

/**
 * Колонки таблиці впорядників: закріплення попереду, а власний гурток прибрано — виховника членство
 * ставить у курінь і зазвичай у жоден гурток, тож ця колонка в них порожня й лише збиває з пантелику.
 */
export function staffColumnsOf(columns: readonly RegistryColumn[]): RegistryColumn[] {
  const assignment = REGISTRY_COLUMNS.find(column => column.id === 'mentoredGroups')!;
  return [assignment, ...columns.filter(column => column.id !== 'groupName' && column.id !== 'mentoredGroups')];
}
